docs/design/uplink-scopes.md

```diff
@@ -0,0 +1,67 @@
+
+There are complications when it comes to restricting to a whitelist of buckets and/or path prefixes inside of those buckets. For example, if a macaroon is limited to only allow access to paths beneath `sj://bucket/a/b`, then it would be impossible to list the root. Additionally, the clients and servers all operate on **encrypted** paths, making it impossible for the server to issue the list and filter, because one would have to reveal the root encryption key for the bucket for the client to be able to decrypt the names.
+
+## Design
+
+Uplink will operate on a set of "scopes". Each scope contains an API key, a satellite url, an optional project key, and a list of bucket shares. In Go,
+
+```go
+type Scope struct {
+    APIKey               string // Or perhaps uplink.APIKey
+    SatelliteURL         string
+    ProjectEncryptionKey []byte
+    BucketShares         []BucketShare
+}
+```
+
+A bucket share contains an encrypted path, the matching unencrypted path, the key beginning at that path, and the bucket the path is contained in. In Go,
+
+```go
+type BucketShare struct {
+    Bucket          string
+    EncryptedPath   string
+    UnencryptedPath string
+    EncryptionKey   []byte
+}
+```
+
+If a client attempts to interact with a bucket, and no `BucketShare` matches, and there is a project key set, a key for that bucket is derived using the bucket name and the project key.
+
+When a client attempts to interact with a bucket, it finds the longest matching unencrypted path, breaking ties by using the later one, and appends the rest of the path, encrypted with the key the bucket share specifies, to the bucket share's encrypted path. For example, if one had this set of `BucketShare` objects:
+
+```go
+paths := []BucketShare{
+    {Bucket: "x", EncryptedPath: "/e1/e2/e3", UnecryptedPath: "/a/b/c", EncryptionKey: "key1"},
+    {Bucket: "x", EncryptedPath: "/e1",       UnecryptedPath: "/a",     EncryptionKey: "key2"},
+    {Bucket: "x", EncryptedPath: "/e4",       UnecryptedPath: "/f",     EncryptionKey: "key3"},
+}
+```
+
+Then if one asked for `sj://x/a/d`, the 2st path would be the longest match and `"key2"` would be used to encrypt `d` and it would be appened to `"/e1"`. Similarly, if one asked for `sj://x/a/b/c/f`, then the 1st path would be the longest match and `"key1"` would be used to encrypt `f` and it would be appened to `"/e1/e2/e3"`. If one asked for `sj://x/f/g`, then `"key3"` would be used. Asking for `sj://x/g` would return not found.
+
+Note that the encrypted path does not need to match the unencrypted path, and this allows for having local views of bucket paths in the actual bucket.
```

Comment:
Could you give some examples here? I am really confused about the relationship between `EncryptedPath` and `UnencryptedPath`. Until now I imagined that when sharing a file with someone we won't reveal its absolute path, not even the bucket it is in with the receiver.

Reply:
I'll add an example about how it doesn't necessarily reveal the actual absolute path.

I don't know if sharing a single file is a supported use case we care about existing in this feature. For example, macaroons only support restricting to a prefix, not to an exact path. Additionally, even if we did care, I don't know if masking the bucket is part of that, and that we can't add it in a backwards compatible way later (we can always add/not set fields to a `BucketShare`).

Reply:
Bucket names aren't encrypted in our paths, so I don't think it would be possible to hide the bucket name effectively when sharing.